enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/log/segmented/Segments.java

```diff
@@ -136,131 +135,96 @@ private synchronized SegmentFile createNext( long prevFileLastIndex, long prevIn
+    synchronized ValueRange<Long,SegmentFile> getForIndex( long logIndex )
     {
         return rangeMap.lookup( logIndex );
     }
 
-    SegmentFile last()
+    synchronized SegmentFile last()
     {
         return rangeMap.last();
     }
 
-    /**
-     * Given a prune index, marks segment files with index less than or equal to this prune index for disposal.
-     * @param pruneIndex The value used to mark segment files for disposal. It can be any legal value (starting from -1
-     * until Long.MAX_VALUE). If it is less than the least prevIndex, this method is effectively a no op and the oldest
-     * segment is returned.
-     * @return The segment file with the least index value not marked for disposal.
-     */
     public synchronized SegmentFile prune( long pruneIndex )
     {
-        Collection<SegmentFile> forDisposal = new ArrayList<>();
-        SegmentFile oldestNotDisposed = collectSegmentsForDisposal( pruneIndex, forDisposal );
+        Iterator<SegmentFile> itr = allSegments.iterator();
+        SegmentFile notDisposed = itr.next(); // we should always leave at least one segment
 
-        log.debug( "Segments marked for pruning: %s", forDisposal );
-        for ( SegmentFile segment : forDisposal )
+        while ( itr.hasNext() )
         {
-            try
+            SegmentFile current = itr.next();
+            if ( current.header().prevFileLastIndex() > pruneIndex )
+            {
+                break;
+            }
+
+            if ( !notDisposed.tryClose() )
             {
-                segment.markForDisposal( this::pruneHandler );
+                break;
             }
-            catch ( DisposedException e )
+
+            log.info( "Pruning %s", notDisposed );
+            if ( !notDisposed.delete() )
             {
-                log.warn( "Already marked for disposal", e );
+                log.error( "Failed to delete %s", notDisposed );
             }
+
+            notDisposed = current;
         }
-        return oldestNotDisposed;
-    }
 
-    private SegmentFile collectSegmentsForDisposal( long pruneIndex, Collection<SegmentFile> forDisposal )
-    {
-        Iterator<SegmentFile> itr = segmentFiles.iterator();
-        SegmentFile prev = itr.next(); // there is always at least one segment
-        SegmentFile oldestNotDisposed = prev;
+        rangeMap.remove( notDisposed.header().prevIndex() + 1 );
 
+        itr = allSegments.iterator();
         while ( itr.hasNext() )
         {
-            SegmentFile segment = itr.next();
-            if ( segment.header().prevFileLastIndex() <= pruneIndex )
-            {
-                forDisposal.add( prev );
-                oldestNotDisposed = segment;
-            }
-            else
+            SegmentFile current = itr.next();
+            if ( current == notDisposed )
             {
                 break;
             }
-            prev = segment;
+            itr.remove();
```

Comment:
this will very likely trigger an arraycopy in the `ArrayList`, I am wondering if it isn't better to use a `LinkedList` for `allSegments`.